### PR TITLE
Adding Controller and Service Unit Tests for EmployeeSalaryDetails

### PR DESCRIPTION
**Pull request by Lucien-vanWyk**
I have added the Unit Tests for both the Controller and Service of EmployeeSalaryDetails. All my tests are passing, and I have added screenshots of the coverage reports of said Unit Tests.

image
image

### CODE REVIEW

File: RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+                               .ThrowsAsync(new Exception("An error occurred while updating employee salary information."));
+
+        var result = await controller.UpdateSalary(employeeSalaryDetailsDto);
+
+        var notFoundResult = Assert.IsType<NotFoundObjectResult>(result);
+        Assert.Equal("An error occurred while updating employee salary information.", notFoundResult.Value);
+    }
+
+    [Fact]
+    public async Task GetAllEmployeeSalariesByEmployeeReturnsOkResultWithList()
```
Comment (LlewellynAnthony) [inline RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs:225]: The naming of your test is confusing, it makes it sound like you're testing a getAll() method. You're only fetching the salaries of a specific employee? Rename the test to something like GetSalariesByEmployeePass().

File: HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+
+            _employeeSalaryDetailsServiceMock.Setup(r => r.GetAllEmployeeSalaries());
+
+            _unitOfWorkMock.Setup(x => x.EmployeeSalaryDetails.GetAll(null)).Returns(Task.FromResult(employeeSalariesDto));
+            var result = await _employeeSalaryDetailsService.GetAllEmployeeSalaries();
+
+            Assert.Equal(2, result.Count);
+            Assert.Equal(employeeSalariesDto, result);
+            Assert.NotNull(result);
+            Assert.IsType<List<EmployeeSalaryDetailsDto>>(result);
```
Comment (MatthewSchoeman) [inline HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs:94]:
```suggestion
            Assert.NotNull(result);
            Assert.Equal(2, result.Count);
            Assert.Equal(employeeSalariesDto, result);
            Assert.IsType<List<EmployeeSalaryDetailsDto>>(result);
```

There is no need to check if the value returned is null after you have already checked if the list has a size of two and if the dto is matching. Rather have the first check that it is not null then check if it contains the expected outcome

File: RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        {
+            Id = 1,
+            EmployeeId = 1,
+            Salary = 2000,
+            MinSalary = 1500,
+            MaxSalary = 3000,
+            Remuneration = 2500,
+            Band = EmployeeSalaryBand.Level1,
+            Contribution = null
+        };
```
Comment (MatthewSchoeman) [inline RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs:97]: It is preferred to store the dto in the test data file so the unit test file can be cleaned up

File: HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs (hunk 2 above)
Comment (Lucien-vanWyk) [reply HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs:94]: Apologies, this was an oversight from my end. 

Your code suggestion and explanation does make sense. It doesn't make sense if we check for null after checking list size😅. I'll commit you suggestion😁

File: RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs (hunk 3 above)
Comment (Lucien-vanWyk) [reply RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs:97]: Makes sense. I'll make use of the EmployeeSalaryDetailsTestData instead.

File: RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs (hunk 1 above)
Comment (Lucien-vanWyk) [reply RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs:225]: You've got a point, and you've got a solid name recommendation. Next commit will fix this.

File: RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+    private readonly Mock<IUnitOfWork> _dbMock;
+    private readonly Mock<IEmployeeSalarayDetailsService> _salaryService;
+    private readonly EmployeeSalaryDetailsDto _employeeSalaryDetailsDto;
+    private readonly List<EmployeeSalaryDetailsDto> _salaries;
+
+    public EmployeeSalaryDetailsControllerUnitTest()
+    {
+        _dbMock = new Mock<IUnitOfWork>();
+        _salaryService = new Mock<IEmployeeSalarayDetailsService>();
+        _employeeController = new EmployeeSalaryDetailsController(_salaryService.Object);
```
[the remark this answers is not in the document]
Comment (Lucien-vanWyk) [reply RR.App.Tests/Controllers/HRIS/EmployeeSalaryDetailsControllerUnitTests.cs]: I've refactored the entire Controller Unit Test file. Just double check to see if our code looks good now.

File: HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+using RR.UnitOfWork.Entities.HRIS;
+using Xunit;
+
+namespace HRIS.Services.Tests.Services
+{
+    public class EmployeeSalaryDetailsServiceUnitTest
+    {
+        private readonly Mock<IUnitOfWork> _unitOfWorkMock;
+        private readonly Mock<IEmployeeService> _employeeServiceMock;
+        private readonly Mock<IEmployeeTypeService> _employeeTypeServiceMock;
```
[the remark this answers is not in the document]
Comment (Lucien-vanWyk) [reply HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs]: Done

File: HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        private readonly Mock<IErrorLoggingService> _errorLoggingServiceMock;
+        private readonly Mock<IEmployeeSalarayDetailsService> _employeeSalaryDetailsServiceMock;
+        private readonly Mock<EmployeeSalaryDetails> _employeeSalaryDetailsMock;
```
[the remark this answers is not in the document]
Comment (Lucien-vanWyk) [reply HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs]: Done

File: HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+            var employeeSalaryDetails = new EmployeeSalaryDetails(EmployeeSalaryDetailsTestData.EmployeeSalaryTest1);
+
+            _unitOfWorkMock.Setup(m => m.EmployeeSalaryDetails.Update(It.IsAny<EmployeeSalaryDetails>()))
+                        .ReturnsAsync(EmployeeSalaryDetailsTestData.EmployeeSalaryTest1);
+            var service = new EmployeeSalaryDetailsService(_unitOfWorkMock.Object, _errorLoggingServiceMock.Object);
+
+            var results = await service.UpdateEmployeeSalary(EmployeeSalaryDetailsTestData.EmployeeSalaryTest1);
+
+            Assert.NotNull(results);
```
[the remark this answers is not in the document]
Comment (Lucien-vanWyk) [reply HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs:114]: Done

File: HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+        public EmployeeSalaryDetailsServiceUnitTest()
+        {
+            _unitOfWorkMock = new Mock<IUnitOfWork>();
+            _errorLoggingServiceMock = new Mock<IErrorLoggingService>();
+            _employeeSalaryDetailsServiceMock = new Mock<IEmployeeSalarayDetailsService>();
+            _employeeSalaryDetailsService = new EmployeeSalaryDetailsService(_unitOfWorkMock.Object, _errorLoggingServiceMock.Object);
+        }
+
+        int employeeId = 1;
```
Comment (tbddorothy) [inline HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs:33]: Please put these variables with the others at the top

File: HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+                Salary = salary
+            };
+            var mockEmployeeSalaryDetailsDbSet = new List<EmployeeSalaryDetails> { employeeSalaryDetails }.AsQueryable().BuildMockDbSet();
+            _unitOfWorkMock.Setup(m => m.EmployeeSalaryDetails.Get(It.IsAny<Expression<Func<EmployeeSalaryDetails, bool>>>()))
+                           .Returns(mockEmployeeSalaryDetailsDbSet.Object);
+
+            var result = await _employeeSalaryDetailsService.GetEmployeeSalary(employeeId);
+
+            Assert.NotNull(result);
+            Assert.Equal(salary, result.Salary);
```
[the remark this answers is not in the document]
Comment (Lucien-vanWyk) [reply HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs:57]: Done. The new code that I've added are the lines bellow the code that's being referenced here

File: HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs (hunk 8 above)
Comment (Lucien-vanWyk) [reply HRIS.Services.Tests/Services/EmployeeSalaryDetailsServiceUnitTests.cs]: My bad, variables have been moved.

**Review by LlewellynAnthony — APPROVED** [review]

**Review by tbddorothy — APPROVED** [review]

**Review by MatthewSchoeman — APPROVED** [review]

**Pull request merged**